src/OVAL/oval_session.c

```diff
@@ -0,0 +1,463 @@
+		return 1;
+	}
+
+	session->res_model = oval_agent_get_results_model(session->sess);
+
+	return 0;
+}
+
+int oval_session_evaluate(struct oval_session *session, char *probe_root, agent_reporter fn, void *arg)
+{
+	__attribute__nonnull__(session);
+
+	if (probe_root) {
+		if (setenv("OSCAP_PROBE_ROOT", probe_root, 1) != 0) {
+			oscap_seterr(OSCAP_EFAMILY_OSCAP, "Failed to set the OSCAP_PROBE_ROOT environment variable.");
+			return 1;
+		}
+	}
+
+	if (oval_session_setup_agent(session) != 0) {
+		return 1;
+	}
+
+	oval_agent_eval_system(session->sess, fn, arg);
+	if (oscap_err()) {
+		return 1;
+	}
+
+	session->res_model = oval_agent_get_results_model(session->sess);
+
+	return 0;
+}
+
+int oval_session_export(struct oval_session *session)
+{
+	__attribute__nonnull__(session);
+
+	struct oval_directives_model *dir_model = NULL;
+	struct oscap_source *result = NULL;		/* OVAL Results */
+	const char *filename = NULL;
+	int ret = 0;
+
+	/* Import OVAL Directives if any */
+	if (session->oval.directives && session->res_model) {
+		/* OVAL Directives can be used only if there are OVAL Resutls and these are
+		 * only available if there is a Results model which mean that either
+		 * evaluation or analyse was performed. */
+		if (dir_model)
+			oval_directives_model_free(dir_model);
+		oval_directives_model_import_source(dir_model, session->oval.directives);
+	}
+
+	/* Get OVAL Results if evaluation or analyse has been done and apply
+	 * directives to them */
+	if (session->res_model && (session->export.results || session->export.report)) {
+		result = oval_results_model_export_source(session->res_model, dir_model, NULL);
+		filename = session->export.results;
+	}
+
+	/* Validate OVAL Results. The 'result' in condition will make sure that there is
+	 * something to validate */
+	if (result && session->validation && session->full_validation) {
+		if (!oval_session_validate(session, result, OSCAP_DOCUMENT_OVAL_RESULTS)) {
+		/* FIXME: if invalid output than add suffix ".invalid" to file name but
+		 * still export and set some error about it ?*/
+		}
+	}
+
+	if (session->export.results && result) {	/* export to XML */
+		if (oscap_source_save_as(result, filename) != 0) {
+			oscap_source_free(result);
+			return 1;
+		}
+	}
+
+	if (session->export.report && result) {	/* export to HTML */
+		char pwd[PATH_MAX];
+
+		if (getcwd(pwd, sizeof(pwd)) == NULL) {
+			oscap_seterr(OSCAP_EFAMILY_GLIBC, "ERROR: %s", strerror(errno));
+			oscap_source_free(result);
+			return 1;
+		}
+
+		/* add params oscap-version & pwd */
+		const char *stdparams[] = { "oscap-version", oscap_get_version(), "pwd", pwd, NULL };
+
+		/* TODO: let the user set the xsl by oval_session_set_report_xsl? */
+		if (oscap_source_apply_xslt_path(result, oval_results_report,
+				session->export.report, stdparams, oscap_path_to_xslt()) == -1) {
+			oscap_source_free(result);
+			return 1;
+		}
+	}
+
+	oscap_source_free(result);
+	if (dir_model)
+		oval_directives_model_free(dir_model);
```

Comment:
Move this ^^^ upper.

Otherwise all the `return 1;` situations will potentially leak `dir_model`